session_jsonl_path: encoding mismatch for paths with dot-prefixed segments (e.g. .worktrees) causes false 'Session ID already in use'

## Summary

`hephaestus/automation/session_naming.py:173-181` (`session_jsonl_path`) builds the on-disk path of a Claude session transcript by replacing only `/` with `-` in the resolved cwd. But Claude Code itself **also** converts each `.` in path segments to `-` when building the projects directory name. The two encodings diverge whenever any path component starts with `.` (e.g. `.worktrees`, `.git`, `.venv`).

The consequence is exactly the bug observed during the 2026-05-30 ecosystem-wide `drive_prs_green.py` run: `transcript.exists()` returns False for a session whose JSONL is actually on disk, the caller goes down the `--session-id` (create) path, and Claude rejects with `Error: Session ID <uuid> is already in use.` — because the session truly does exist server-side, just at a different on-disk encoding than the code is probing.

## Repro (verified live 2026-05-30)

```bash
$ ls /home/mvillmow/.claude/projects/-home-mvillmow-Projects-ProjectOdyssey-build--worktrees-issue-5451/
9ede78c6-262c-5931-b832-3897c3faf71f.jsonl

$ python3 -c "
from pathlib import Path
p = Path('/home/mvillmow/Projects/ProjectOdyssey/build/.worktrees/issue-5451')
encoded = str(p.resolve()).replace('/', '-')
print('hephaestus encodes to:', encoded)
print('claude actually uses :', '-home-mvillmow-Projects-ProjectOdyssey-build--worktrees-issue-5451')
print('match:', encoded == '-home-mvillmow-Projects-ProjectOdyssey-build--worktrees-issue-5451')
"
hephaestus encodes to: -home-mvillmow-Projects-ProjectOdyssey-build-.worktrees-issue-5451
claude actually uses : -home-mvillmow-Projects-ProjectOdyssey-build--worktrees-issue-5451
match: False
```

The mismatch is `-build-.worktrees-` (hephaestus) vs `-build--worktrees-` (Claude — `.` rewritten to `-`).

## Resulting failure in the wild

From the live ecosystem run (`~/drive-prs-green-logs/.../ProjectOdyssey.log`):

```
2026-05-30 15:45:20 [ERROR] hephaestus.automation.ci_driver: Issue #5451: Claude CI fix session returned exit code 1: Error: Session ID 9ede78c6-262c-5931-b832-3897c3faf71f is already in use.
2026-05-30 15:45:33 [ERROR] hephaestus.automation.ci_driver: Issue #5391: Claude CI fix session returned exit code 1: Error: Session ID e2089c9c-e10f-54dd-9836-ee83032c0fbe is already in use.
2026-05-30 15:46:09 [ERROR] hephaestus.automation.ci_driver: Issue #5450: Claude CI fix session returned exit code 1: Error: Session ID c25b52cd-eb74-56a6-b8db-19fcd6c36bd9 is already in use.
2026-05-30 15:46:19 [ERROR] hephaestus.automation.ci_driver: Issue #5449: Claude CI fix session returned exit code 1: Error: Session ID d72c5300-a73b-508a-8112-a141444af160 is already in use.
```

Every issue's CI fix failed on the first attempt because the worktree cwd contains `.worktrees`. Same root cause for every entry.

## Fix

In `hephaestus/automation/session_naming.py:173-181`, replace the encoding step:

```python
def session_jsonl_path(uuid_str: str, cwd: Path) -> Path:
    """..."""
    # Claude Code rewrites BOTH `/` and `.` to `-` when building the
    # projects directory name. Match that encoding exactly so paths
    # containing dot-prefixed segments (.worktrees, .git, .venv) probe
    # the right on-disk transcript.
    resolved = str(cwd.resolve())
    encoded = resolved.replace("/", "-").replace(".", "-")
    return Path.home() / ".claude" / "projects" / encoded / f"{uuid_str}.jsonl"
```

Order matters in principle but not here: neither `/` produces `.` nor vice versa, so the two `replace` calls commute. Document the encoding rule in the docstring.

## Hardening (recommended, not strictly required for this bug)

Even with the encoding fix, the `--session-id` (create) path at `hephaestus/automation/claude_invoke.py:201` has no fallback when the CLI reports the session is already in use. Add a fallback symmetric to the existing resume-failure recreate path:

```python
try:
    return _run(create=True).stdout, sid
except subprocess.CalledProcessError as exc:
    stderr = exc.stderr or ""
    if "already in use" in stderr.lower():
        # The probe missed a session that exists server-side; resume it.
        logger.warning("claude --session-id %s rejected as already in use; resuming instead", sid)
        return _run(create=False).stdout, sid
    raise
```

This is defense in depth — protects against encoding drift, race conditions where another process created the session between probe and invoke, and CLI-behavior changes.

## Acceptance criteria

- `session_jsonl_path(uuid, Path('/x/.worktrees/y'))` returns `~/.claude/projects/-x--worktrees-y/<uuid>.jsonl` (not `…-x-.worktrees-y…`).
- A re-run of `drive_prs_green.py` against a previously-driven issue (whose JSONL exists on disk) goes down the `--resume` path, not the `--session-id` create path. No `Session ID … is already in use` errors.
- New unit tests in `tests/unit/automation/test_session_naming.py`:
  - cwd with one `.`-prefixed segment → encoded with `--` at that join (no `.`).
  - cwd with multiple `.`-prefixed segments (e.g. `/x/.venv/.tox/y`) → all `.` → `-`.
  - cwd with `.` mid-segment (e.g. `.../v1.2.3/...`) → ALL dots rewritten (matches Claude's actual behavior; document the unlikely-edge-case in the docstring).
- Add the hardening fallback in `claude_invoke.py` (see above), with a unit test that simulates `CalledProcessError` with `already in use` and asserts the function falls through to `--resume`.

## Out of scope

- Investigating whether Claude Code's encoding has *additional* rules (e.g. multiple consecutive `.` collapsing, Unicode normalization). The current cases (`.worktrees`, `.git`, `.venv`) are the immediate need.

--- a/hephaestus/automation/claude_invoke.py
+++ b/hephaestus/automation/claude_invoke.py
@@ -178,7 +178,21 @@ def invoke_claude_with_session(  # noqa: C901  # state machine: argv assembly + 
         )
 
     if not should_resume:
-        return _run(create=True).stdout, sid
+        try:
+            return _run(create=True).stdout, sid
+        except subprocess.CalledProcessError as exc:
+            # Defense in depth (#822): the probe says no transcript exists
+            # but the CLI says the session is already registered. Encoding
+            # drift between hephaestus and the CLI can desync the probe;
+            # treat the rejection as proof the session exists and resume it.
+            stderr = (exc.stderr or "") + (exc.stdout or "")
+            if "already in use" in stderr.lower():
+                logger.warning(
+                    "claude --session-id %s rejected as already in use; resuming instead",
+                    sid,
+                )
+                return _run(create=False).stdout, sid
+            raise
 
     try:
         return _run(create=False).stdout, sid

--- a/hephaestus/automation/session_naming.py
+++ b/hephaestus/automation/session_naming.py
@@ -173,9 +173,13 @@ def current_trunk_githash(repo_path: Path | None = None) -> str:
 def session_jsonl_path(uuid_str: str, cwd: Path) -> Path:
     """Return the path where Claude Code persists a session's transcript.
 
-    Claude encodes the cwd into the directory name by replacing each ``/``
-    with ``-``. Used to detect whether a session already exists (and thus
-    should be resumed) or needs to be created.
+    Claude encodes the cwd into the projects-directory name by replacing
+    BOTH ``/`` and ``.`` with ``-``. Probing only ``/`` (as a prior version
+    of this helper did) misses every cwd containing a dot-prefixed segment
+    like ``.worktrees``, ``.git``, or ``.venv``: ``transcript.exists()``
+    returns False even though the JSONL is on disk, the caller goes down
+    the ``--session-id`` create path, and the CLI rejects with ``Session ID
+    <uuid> is already in use``. (#822)
     """
-    encoded = str(cwd.resolve()).replace("/", "-")
+    encoded = str(cwd.resolve()).replace("/", "-").replace(".", "-")
     return Path.home() / ".claude" / "projects" / encoded / f"{uuid_str}.jsonl"

--- a/tests/unit/automation/test_claude_invoke_session.py
+++ b/tests/unit/automation/test_claude_invoke_session.py
@@ -217,10 +217,10 @@ class TestSessionExpiredFallback:
         assert m.call_count == 1
 
     def test_create_already_in_use_falls_back_to_resume(self, fake_home: Path) -> None:
-        """When the CLI rejects --session-id with 'already in use', fall back to --resume. (#822)
+        """Fall back to --resume when --session-id is rejected as already-in-use.
 
         Defends against encoding drift between hephaestus's transcript probe
-        and the Claude CLI's actual session storage.
+        and the Claude CLI's actual session storage. (#822)
         """
         cwd = fake_home / "work"
         cwd.mkdir()

--- a/tests/unit/automation/test_claude_invoke_session.py
+++ b/tests/unit/automation/test_claude_invoke_session.py
@@ -194,7 +194,7 @@ class TestSessionExpiredFallback:
         assert "--session-id" in _argv(m.call_args_list[1])
 
     def test_create_failure_propagates(self, fake_home: Path) -> None:
-        """A first-call (--session-id) failure is not retried."""
+        """A first-call (--session-id) failure for an unrelated reason is not retried."""
         cwd = fake_home / "work"
         cwd.mkdir()
         other = subprocess.CalledProcessError(
@@ -215,6 +215,43 @@ class TestSessionExpiredFallback:
                     cwd=cwd,
                 )
         assert m.call_count == 1
+
+    def test_create_already_in_use_falls_back_to_resume(self, fake_home: Path) -> None:
+        """When the CLI rejects --session-id with 'already in use', fall back to --resume. (#822)
+
+        Defends against encoding drift between hephaestus's transcript probe
+        and the Claude CLI's actual session storage.
+        """
+        cwd = fake_home / "work"
+        cwd.mkdir()
+        already = subprocess.CalledProcessError(
+            returncode=1,
+            cmd=["claude"],
+            output="",
+            stderr="Error: Session ID abc is already in use.",
+        )
+        ok = subprocess.CompletedProcess(
+            args=["claude"], returncode=0, stdout="resumed-ok", stderr=""
+        )
+        with patch(
+            "hephaestus.automation.claude_invoke.subprocess.run",
+            side_effect=[already, ok],
+        ) as m:
+            stdout, _ = invoke_claude_with_session(
+                repo="R",
+                issue=1,
+                agent=AGENT_PLANNER,
+                githash="x",
+                prompt="hi",
+                model="sonnet",
+                cwd=cwd,
+            )
+        assert stdout == "resumed-ok"
+        assert m.call_count == 2
+        first_argv = m.call_args_list[0][0][0]
+        second_argv = m.call_args_list[1][0][0]
+        assert "--session-id" in first_argv
+        assert "--resume" in second_argv
 
 
 class TestArgvAssembly:

--- a/tests/unit/automation/test_session_naming.py
+++ b/tests/unit/automation/test_session_naming.py
@@ -192,8 +192,9 @@ class TestSessionJsonlPath:
     def test_dot_prefixed_segment_encodes_dot_as_dash(
         self, tmp_path: Path, monkeypatch: pytest.MonkeyPatch
     ) -> None:
-        """A `.worktrees` segment must yield `--worktrees-`, matching the
-        Claude CLI's actual on-disk encoding. (#822)
+        """Encode a `.worktrees` segment as `--worktrees-`, not `-.worktrees-`.
+
+        Matches the Claude CLI's actual on-disk encoding. (#822)
         """
         monkeypatch.setenv("HOME", str(tmp_path))
         target = tmp_path / "build" / ".worktrees" / "issue-5451"
@@ -216,8 +217,9 @@ class TestSessionJsonlPath:
     def test_mid_segment_dot_also_rewritten(
         self, tmp_path: Path, monkeypatch: pytest.MonkeyPatch
     ) -> None:
-        """Mid-segment dots (e.g. `v1.2.3`) are rewritten too — matches the
-        CLI's flat `.` -> `-` rule. Documented edge case, not a bug.
+        """Rewrite mid-segment dots (e.g. `v1.2.3`) too.
+
+        Matches the CLI's flat `.` -> `-` rule. Documented edge case, not a bug.
         """
         monkeypatch.setenv("HOME", str(tmp_path))
         target = tmp_path / "release" / "v1.2.3" / "build"

--- a/tests/unit/automation/test_session_naming.py
+++ b/tests/unit/automation/test_session_naming.py
@@ -176,13 +176,11 @@ class TestSessionJsonlPath:
 
     def test_path_encoding(self, tmp_path: Path, monkeypatch: pytest.MonkeyPatch) -> None:
         monkeypatch.setenv("HOME", str(tmp_path))
-        # Path.home() reads $HOME at call time, so no reimport needed.
         target = tmp_path / "Projects" / "Foo"
         target.mkdir(parents=True)
         p = session_jsonl_path("abc-uuid", target)
         assert p.name == "abc-uuid.jsonl"
-        # The encoded segment is the resolved path with `/` -> `-`.
-        encoded = str(target.resolve()).replace("/", "-")
+        encoded = str(target.resolve()).replace("/", "-").replace(".", "-")
         assert encoded in str(p)
         assert p.parent.parent == tmp_path / ".claude" / "projects"
 
@@ -190,6 +188,43 @@ class TestSessionJsonlPath:
         sid = session_uuid("R", 1, AGENT_IMPLEMENTER, "abc1234")
         p = session_jsonl_path(sid, tmp_path)
         assert p.name == f"{sid}.jsonl"
+
+    def test_dot_prefixed_segment_encodes_dot_as_dash(
+        self, tmp_path: Path, monkeypatch: pytest.MonkeyPatch
+    ) -> None:
+        """A `.worktrees` segment must yield `--worktrees-`, matching the
+        Claude CLI's actual on-disk encoding. (#822)
+        """
+        monkeypatch.setenv("HOME", str(tmp_path))
+        target = tmp_path / "build" / ".worktrees" / "issue-5451"
+        target.mkdir(parents=True)
+        p = session_jsonl_path("u", target)
+        # `/.worktrees/` becomes `--worktrees-`, NOT `-.worktrees-`.
+        assert "--worktrees-issue-5451" in str(p)
+        assert "-.worktrees-" not in str(p)
+
+    def test_multiple_dot_segments_all_rewritten(
+        self, tmp_path: Path, monkeypatch: pytest.MonkeyPatch
+    ) -> None:
+        monkeypatch.setenv("HOME", str(tmp_path))
+        target = tmp_path / ".venv" / ".tox" / "y"
+        target.mkdir(parents=True)
+        p = session_jsonl_path("u", target)
+        # Both leading dots get rewritten.
+        assert "." not in p.parent.name
+
+    def test_mid_segment_dot_also_rewritten(
+        self, tmp_path: Path, monkeypatch: pytest.MonkeyPatch
+    ) -> None:
+        """Mid-segment dots (e.g. `v1.2.3`) are rewritten too — matches the
+        CLI's flat `.` -> `-` rule. Documented edge case, not a bug.
+        """
+        monkeypatch.setenv("HOME", str(tmp_path))
+        target = tmp_path / "release" / "v1.2.3" / "build"
+        target.mkdir(parents=True)
+        p = session_jsonl_path("u", target)
+        assert "v1-2-3" in p.parent.name
+        assert "v1.2.3" not in p.parent.name
 
 
 class TestCurrentTrunkGithash:

